File: P/Project_1/Go_Implementation/Feistal_Network/feistalN.go
package main

func permutation(charToPermutate byte) byte {
	return charToPermutate
}

func feistalIter(firstSlice []byte, secondSlice []byte) ([]byte, []byte) {
	BLOCK_SIZE := 8

	// Iterate over each char
	for i := 0; i < BLOCK_SIZE/2; i++ {
		//fmt.Printf("firstSlice: %d, secondSlice: %d\n", i, firstSlice[0], secondSlice[0]) // DEBUG

		firstChar := firstSlice[i]
		secondChar := secondSlice[i]

		// Apply permutation
		secondChar = permutation(secondChar)

		// Apply XOR Li with fi and switch position between slices
		firstSlice[i] = secondChar
		secondSlice[i] = firstChar ^ secondChar
		//fmt.Printf("Iter: %d, firstSlice: %d, secondSlice: %d\n", i, firstSlice[0], secondSlice[0]); // DEBUG

	}

	return firstSlice, secondSlice
}

func main() {
	NUM_ITERATIONS := 16

	firstSlice := []byte("IIIIIIII")
	secondSlice := []byte("OOOOOOOO")

	for i := 0; i < NUM_ITERATIONS; i++ {
		//fmt.Printf("Iter: %d\nFirstSlice: %d, secondSlice: %d\n", i, firstSlice[0], secondSlice[0]) // DEBUG
		firstSlice, secondSlice = feistalIter(firstSlice, secondSlice)
	}

	/** // Check for unwind of the feistal Network

	for i := 0; i < NUM_ITERATIONS; i++ {
		//printf("Iter: %d\nFirstSlice: %d, secondSlice: %d\n", i, firstSlice[0], secondSlice[0]); // DEBUG
		feistalIter(secondSlice, firstSlice)
	}

	for i := 0; i < 9; i += 1 {
		fmt.Printf("FirstSlice: %c, secondSlice: %c\n", firstSlice[0], secondSlice[0]) // DEBUG
	}
	**/
}

File: P/Project_1/C_Implementation/Feistal_Network/feistalN.c
#include <stdio.h>

int permutation(char charToPermutate) {
    return 0;
}

int feistalIter(char firstSlice [], char secondSlice []) {
    int BLOCK_SIZE = 8;

    // Iterate over each char
    for( int i = 0 ; i < BLOCK_SIZE/2; i += 1 ){
        //printf("firstSlice: %c, secondSlice: %c\n", firstSlice[i], secondSlice[i]); // DEBUG

        char firstChar = firstSlice[i];
        char secondChar = secondSlice[i];

        // Apply permutation
        permutation(secondChar);

        // Apply XOR Li with fi and switch position between slices
        firstSlice[i] = secondChar;
        secondSlice[i] = firstChar ^ secondChar;
        //printf("Iter: %d, firstSlice: %d, secondSlice: %d\n", i, firstSlice[0], secondSlice[0]); // DEBUG
    }

    return 0;
}

int main() {
    int NUM_ITERATIONS = 16;

    char firstSlice[] = "IIIIIIII";
    char secondSlice[] = "OOOOOOOO";

    for( int i = 0 ; i < NUM_ITERATIONS; i += 1 ){
        //printf("Iter: %d\nFirstSlice: %d, secondSlice: %d\n", i, firstSlice[0], secondSlice[0]); // DEBUG
        feistalIter(firstSlice, secondSlice);
    }

    /** Check for unwind of the feistal Network

    for( int i = 0 ; i < NUM_ITERATIONS; i += 1 ){
        //printf("Iter: %d\nFirstSlice: %d, secondSlice: %d\n", i, firstSlice[0], secondSlice[0]); // DEBUG
        feistalIter(secondSlice, firstSlice);
    }

    for( int i = 0 ; i < 9; i += 1 ){
        printf("FirstSlice: %c, secondSlice: %c\n", firstSlice[0], secondSlice[0]); // DEBUG
    }
    **/

    return 0;
}